rclpy/src/rclpy/_rclpy.c

```diff
@@ -1599,6 +1599,72 @@ rclpy_publish(PyObject * Py_UNUSED(self), PyObject * args)
   Py_RETURN_NONE;
 }
 
+/// Publish a serialized message
+/**
+ * Raises ValueError if pypublisher is not a publisher capsule
+ * Raises RuntimeError if the message cannot be published
+ *
+ * \param[in] pypublisher Capsule pointing to the publisher
+ * \param[in] pymsg serialized message to send
+ * \return NULL
+ */
+
+static PyObject *
+rclpy_publish_raw(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pypublisher;
+  PyBytesObject * pymsg;
+  char * message;
+  Py_ssize_t pymsg_size;
+
+  if (!PyArg_ParseTuple(args, "OO", &pypublisher, &pymsg)) {
+    return NULL;
+  }
+
+  int err = PyBytes_AsStringAndSize(pymsg, &message, &pymsg_size);
+
+  if (err == -1) {
+    PyErr_Format(RCLError, "Failed to fetch buffer contents: %d", err);
+    return NULL;
+  }
+
+  rcl_serialized_message_t msg = rmw_get_zero_initialized_serialized_message();
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcl_ret_t ret = rmw_serialized_message_init(&msg, (size_t)pymsg_size, &allocator);
+
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(
+      RCLError,
+      "Failed to initialize message: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
+    if (r_fini != RMW_RET_OK) {
+      PyErr_Format(RCLError, "Failed to deallocate message buffer: %d", r_fini);
+    }
+    return NULL;
+  }
+
+  msg.buffer = message;
+
+  rclpy_publisher_t * pub = (rclpy_publisher_t *)PyCapsule_GetPointer(
+    pypublisher, "rclpy_publisher_t");
+
+  if (!pub) {
+    return NULL;
+  }
+
+  ret = rcl_publish_serialized_message(&(pub->publisher), &msg, NULL);
+
```

Comment:
Here, we need to finalize the message we allocated previously.

Reply:
finalizing the message at the end causes a heap corruption error while testing. Is it because the message buffer is already de-allocated in `rcl_publish_serialized_message` and we're trying to free it again? 

```c
  rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
  if (r_fini != RMW_RET_OK) {
    PyErr_Format(RCLError, "Failed to deallocate message buffer: %d", r_fini);
    return NULL;
  }
```
```
64.672s] 17: ..\..\src\ros2\rclpy\rclpy\test\test_messages.py .Windows fatal exception: code 0xc0000374
[64.672s] 17: 
[64.672s] 17: Current thread 0x00005c30 (most recent call first):
[64.672s] 17:   File "E:\Projects\RoS\src\ros2\rclpy\rclpy\rclpy\publisher.py", line 71 in publish
[64.672s] 17:   File "E:\Projects\RoS\src\ros2\rclpy\rclpy\test\test_messages.py", line 69 in test_serialized_publish
[64.672s] 17:   File "E:\Software\Python36\lib\unittest\case.py", line 605 in run
[64.672s] 17:   File "E:\Software\Python36\lib\unittest\case.py", line 653 in __call__
[64.672s] 17:   File "E:\Software\Python36\lib\site-packages\_pytest\unittest.py", line 207 in runtest
[64.672s] 17:   File "E:\Software\Python36\lib\site-packages\_pytest\runner.py", line 134 in pytest_runtest_call
[64.672s] 17:   File "E:\Software\Python36\lib\site-packages\pluggy\callers.py", line 187 in _multicall
[64.672s] 17:   File "E:\Software\Python36\lib\site-packages\pluggy\manager.py", line 87 in <lambda>
[64.672s] 17:   File "E:\Software\Python36\lib\site-packages\pluggy\manager.py", line 93 in _hookexec
[64.672s] 17:   File "E:\Software\Python36\lib\site-packages\pluggy\hooks.py", line 286 in __call__
```

Reply:
`rcl_publish_serialized_message` is not deallocating the the serialized message. I don't know why you would experience a heap corruption. Where is the snippet you are referring to?

Reply:
See my comment for why we're seeing a segfault: https://github.com/ros2/rclpy/pull/509#discussion_r391787850